Transparent proxy injection

rmars opened the pull request:
Transparently proxy all outgoing requests to a daemonset linkerd via iptables rules (supplied by an init container).

This only looks big because of /vendor. Relevant code is in https://github.com/linkerd/linkerd-inject/commit/ce4887e4216469c0ac1bec2b3c1174ae3a43bfde.

This is based on istioctl's [proxy injection](https://github.com/istio/pilot/blob/master/doc/proxy-injection.md) (they run a sidecar proxy and do transparent proxying via iptables rules).

- Adds a Dockerfile to create an init-container image that we can use to apply iptables rules to the pod. These rules forward outgoing pod traffic to a daemonset linkerd
- Add prepare_proxy.sh which sets up these rules
- Add an inject.go which will add the initContainer to the user's resource file 

All this is based heavily on istio's existing code and ideally this would go with that code somewhere, but I think it is being moved around right now. Would be nice to delete our inject.go in favour of something in that repo after it has settled.

Testing:
Ran inject.go against yamls in https://github.com/istio/pilot/tree/master/platform/kube/testdata.
Used our hello world example, ran it in minikube and in a 3 node GKE cluster.

Notes:
* When the services call each other, the service names need to be resolvable by kubedns so that the packet has an IP and can be routed (hence renaming world-v1 to world).
* I ran into a go dependency mess, see https://github.com/kubernetes/client-go/blob/master/INSTALL.md

Fixes linkerd#1415

rmars committed:
Add transparent proxy injection, based on istioctl kube-inject

* Add script which generates iptables rules to forward traffic to a daemonset linkerd
* Add Dockerfile to build this initContainer
* Add go script to inject this initContainer into the user's config

rmars committed:
Add vendored deps via dep

Per https://github.com/kubernetes/client-go/blob/master/INSTALL.md

klingerf commented on README.md:
````diff
@@ -1 +1,51 @@
+```
+
+It is based on Istio's method of
+[injecting sidecars](https://github.com/istio/pilot/blob/master/doc/proxy-injection.md),
````
Super minor, but since this code is in flux right now, maybe consider pinning this to a specific release tag, rather than master. How about `https://github.com/istio/pilot/blob/stable-a632657/doc/proxy-injection.md`?

rmars replied:
ooh, yeah that's a good idea! thanks!

klingerf commented on README.md:
```diff
@@ -1 +1,51 @@
+more directly, but this seems to be in transit right now. This `prepare_proxy.sh`
+sets up iptables rules for transparently proxying requests to a Daemonset linkerd
+(rather than a sidecar proxy, which Istio
+[currently uses](https://github.com/istio/pilot/blob/master/docker/prepare_proxy.sh)).
```
Same comment goes for here (and the rest of the github.com/istio/pilot links in this diff): `https://github.com/istio/pilot/blob/stable-a632657/docker/prepare_proxy.sh`

klingerf reviewed (approved):
⭐️  This looks good to me -- nice work!

rmars committed:
pin all istio/pilot github links to the working istioctl branch

esbie commented on example/helloworld.yaml:
```diff
@@ -0,0 +1,88 @@
+---
```
maybe name this hello-world.yml to match the linkerd-examples examples

rmars replied:
oops, yes, thanks!

esbie commented on inject.go:
```diff
@@ -0,0 +1,235 @@
+					},
+				},
+			},
+		},
```
Could you add this env section to the comment at the top of the file?

I'm a little concerned that this means this doesn't work in minikube? It's not a blocker for this initial PR, but I definitely am interested in finding a way to support minikube.

rmars replied:
Yeahhhh agreed. Since minikube has one node, maybe we could use the l5d service VIP in the iptables rules rather than the host ip (which we can't get in minikube)?

So I can get this to run in minikube by doing the following, though it's a bit much to ask a user to do this:

Modify prepare_proxy.sh to use DNAT to l5d service VIP:
```
iptables -t nat -A OUTPUT -p tcp -j DNAT --to ${L5D_SERVICE_HOST}:${LINKERD_PORT}                  -m comment --comment "istio/dnat-to-daemonset-l5d"
```

Build a linkerd/istio-init docker image locally and use that.
Use inject.go or manually insert the initContainer section in your pod spec. Make sure to remove` imagePullPolicy: Always` from this section though.

Deploy your modified yaml. Then:
```
L5D_PORT=$(kubectl get svc l5d -o jsonpath="{.spec.ports[0].nodePort}")
curl -v $(minikube ip):$L5D_PORT -H "Host: hello"
```

I was just talking to @adleong about this and maybe we can add a flag to `inject.go` so that people can run in minikube without having to build their own image for this.

rmars replied:
Okay added a `-runInMinikube` in 2a128ea

esbie commented on docker/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+# Build linkerd inject image
+#
+#  docker build -t buoyantio/istio-init:v1 .
+#  docker push buoyantio/istio-init:v1
```
How about putting this at `linkerd/istio-init` instead?

esbie reviewed (commented):
This is looking very nice ^_^

rmars committed:
Address esbies comments

Add env section to comment in inject.go

rename helloworld.yaml to hello-world.yml

use a linkerd/istio-init image name

rmars committed:
Name the init container init-linkerd, s/buoyantio/linkerd in image name

rmars committed:
Add iptables rules and inject.go flag for running in minikube

rmars committed:
Tweak readmes

esbie commented on docker/prepare_proxy.sh:
```diff
@@ -0,0 +1,56 @@
+
+if [ "$RUN_IN_MINIKUBE" = true ]; then
+  # Forward traffic to the linkerd service VIP on the single minikube node
+  iptables -t nat -A OUTPUT -p tcp -j DNAT --to ${L5D_SERVICE_HOST}:${LINKERD_PORT}         -m comment --comment "istio/dnat-to-minikube-l5d"
```
I'm a little confused, where is L5D_SERVICE_HOST set?

rmars replied:
Kubernetes creates these env variables (https://kubernetes.io/docs/concepts/containers/container-environment-variables/)
This does assume we've named the linkerd service "l5d" - so I should probably explicitly pass in the service name

esbie commented:
`-runInMinikube` is straightforward but not descriptive. Would prefer `--use-vip` or `--single-node` or something, because minikube isn't the only kubernetes installation without downward api access. In fact I was just helping someone with that kind of setup yesterday in https://github.com/linkerd/linkerd/issues/1464

adleong commented:
or maybe `--l5d-host`

rmars committed:
Rename runInMinikube to useServiceVip, add configurable l5d service name

rmars committed:
Tweak readme

rmars committed:
Add github link

rmars committed:
Add TODO

esbie reviewed (approved):
the changes look great! 💡 🎥  🎬
